draft-ietf-httpbis-priority.md

```diff
@@ -185,136 +185,72 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
 
 Unknown parameters, parameters with out-of-range values or values of unexpected
 types MUST be ignored.
 
-## urgency
+## Urgency
 
-The urgency(`u`) parameter takes an integer between 0 and 7, in descending order of
-priority, as shown below:
+The urgency parameter (`u`) takes an integer between 0 and 7, in descending
+order of priority.
 
-| Urgency         | Definition                        |
-|----------------:|:----------------------------------|
-|               0 | prerequisite ({{prerequisite}})   |
-|               1 | default ({{default}})             |
-| between 2 and 6 | supplementary ({{supplementary}}) |
-|               7 | background ({{background}})       |
-{: #urgencies title="Urgencies"}
+The value is encoded as an sh-integer. The default value is 3.
 
-The value is encoded as an sh-integer. The default value is 1.
-
-A server SHOULD transmit HTTP responses in the order of their urgency values.
-The lower the value, the higher the precedence.
+This parameter indicates the sender's recommendation, based on the expectation
+that the server would transmit HTTP responses in the order of their urgency
+values if possible. The smaller the value, the higher the precedence.
 
 The following example shows a request for a CSS file with the urgency set to
-`0`:
+`2`:
 
 ~~~ example
 :method = GET
 :scheme = https
 :authority = example.net
 :path = /style.css
-priority = u=0
+priority = u=2
 ~~~
 
-The definition of the urgencies and their expected use-case are described below.
-Endpoints SHOULD respect the definition of the values when assigning urgencies.
-
-### prerequisite
-
-The prerequisite urgency (value 0) indicates that the response prevents other
-responses with an urgency of prerequisite or default from being used until it
-is fully transmitted.
-
-For example, use of an external stylesheet can block a web browser from
-rendering the HTML. In such case, the stylesheet is given the prerequisite
-urgency.
-
-### default
-
-The default urgency (value 1) indicates a response that is to be used as it is
-delivered to the client, but one that does not block other responses from being
-used.
-
-For example, when a user using a web browser navigates to a new HTML document,
-the request for that HTML is given the default urgency.  When that HTML document
-uses a custom font, the request for that custom font SHOULD also be given the
-default urgency.  This is because the availability of the custom font is likely
-a precondition for the user to use that portion of the HTML document, which is
-to be rendered by that font.
-
-### supplementary
-
-The supplementary urgencies (values 2 to 6) indicate a response that is helpful
-to the client using a composition of responses, even though the response itself
-is not mandatory for using those responses.
+A client that fetches a document that likely consists of multiple HTTP resources
+(e.g., HTML) SHOULD assign the default urgency level to the main resource.  This
```

Comment:
Sorry I missed the meeting. Can you explain the rationale for this? Since there is no resource more urgent than the main resource, this effectively reserves priorities 0-2 for the server.

Reply:
Sorry if you felt left alone. I have written down my understanding of the outcome of the meeting at https://github.com/httpwg/http-extensions/issues/1026#issuecomment-582776695, I think we might want to consider discussing there.

Reply:
No worries and thanks for the context. The examples make sense. I think it's fine to say that 7 should be reserved for background downloads and it's fine to say that the main HTML should use level X, for some X, but I'm not sure why X=3 was chosen. It seems like this could work just as well with X=0 or X=1.

Note that Chrome (and WebKit/Safari?) use 6 priority levels currently. With X = 3, there are 5 levels available which means that Chrome (and WebKit/Safari?) need to fold two levels into one. This is not a dealbreaker but IMO there should at least be a good reason why levels 0-2 need to be reserved for the server when clients have a demonstrated use for 6 of 8 levels. Ideally someone from Firefox (and Safari) should chime in as well.

It's also worth mentioning that this pull request doesn't have enough information to do everything that [this comment](https://github.com/httpwg/http-extensions/issues/1026#issuecomment-582776695) wants to do. For example: "a server might demote them to the same level as asynchronously fetched JavaScript files, or to somewhere even below, but above background downloads." It's not possible to demote to the same level as async JS, because the level of async JS is not specified. Also, it's not trivial to set a priority "above background downloads" because the pull request merely states that level 7 is reserved for background downloads, not anything about level 6 (perhaps some client has many background downloads and wants to partition those into two levels). I wouldn't change this pull request though -- this pull request seems like the right level of detail.

To summarize, this looks good but I would suggest changing the level of mainframe requests to 0 or 1.

Reply:
Thank you for sharing your thoughts.

Regarding background, you are right! I hadn't thought about that, but the new design does not forbid a client from using level(s) above 7 for background tasks too. That said, I agree with you that we do not need to be more strict.

Regarding the default level, based on your comments, I am going to change it to 1. I think it would be a good idea to provide one level above default, considering the fact that Firefox prioritizes JavaScript and CSS files above HTML. But I do not think we need more than one.

Then, if people prefer having symmetry (i.e. having same number of urgency levels above and below the default), we can consider increasing the number of levels to 10 (or maybe 12) so that there would be sufficient number of levels below the default.

Reply:
Done in bc04fca.